include/dali/imgcodec/image_decoder_interfaces.h

```diff
@@ -75,10 +75,35 @@ struct ROI {
   }
 };
 
-struct DecodeResult {
-  bool success;
-  std::exception_ptr exception;
-};
+template <typename OutShape>
+void OutputShape(OutShape &out_shape,
+                 const ImageInfo &info, const DecodeParams &params, const ROI &roi) {
+  int ndim = info.shape.sample_dim();
+  resize_if_possible(out_shape, ndim);
+  assert(size(out_shape) == ndim);  // check the size, in case we couldn't resize
+
+  int spatial_ndim = ndim - 1;
+  int in_channel_dim = ndim - 1;
+  int num_channels = NumberOfChannels(params.format, info.shape[in_channel_dim]);
+  int out_channel_dim = params.planar ? 0 : ndim - 1;
+
+  out_shape[out_channel_dim] = num_channels;
+
+  for (int d = 0; d < spatial_ndim; d++) {
+    int in_d = d;
+    if (params.use_orientation && ((info.orientation.rotate / 90) & 1)) {
+      if (spatial_ndim != 2)
+        throw std::logic_error("Orientation only applied to 2D images.");
+      in_d = 1 - d;
+    }
```

Comment:
Btw. it seems you could have `bool flip_dims = params.use_orientation && ((info.orientation.rotate / 90) & 1)` and the spatial_dim == 2 check outside the loop, right?

Reply:
I guess...